[dynamo] Fix bug in tensor.item fake tensor propogation

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91668

When we run the node with fake value for tensor.item, it would previously error because the utility method doesn't know how to handle placeholder node. The tensor we are calling item can be input from user will be placeholder in the graph. 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

tugsbayasgalan committed:
[dynamo] Fix bug in tensor.item fake tensor propogation

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/utils.py:
```diff
@@ -1057,6 +1057,9 @@ def run_node(output_graph, node, args, kwargs, nnmodule):
         elif op == "get_attr":
             return output_graph.get_submodule(node.target)
+        elif op == "placeholder":
+            if "example_value" in node.meta:
+                return node.meta["example_value"]
```
This now returns none if there no example_value in node.meta, can we keep it going to RuntimeError in these cases? or, if you prefer, assert that we have an example_value for placeholder.

tugsbayasgalan replied:
I think it is fine because it would raise an Exception as we don't return in the try/except block.

tugsbayasgalan replied:
https://github.com/pytorch/pytorch/blob/6bf0e3b697ce688bc8325440dea3b51fea571c3d/torch/_dynamo/utils.py#L1060

voznesenskym commented on torch/_dynamo/utils.py:
```diff
@@ -1057,6 +1057,9 @@ def run_node(output_graph, node, args, kwargs, nnmodule):
             return nnmodule(*args, **kwargs)
         elif op == "get_attr":
             return output_graph.get_submodule(node.target)
+        elif op == "placeholder":
```
j/w when does run_node get called on a placeholder?

tugsbayasgalan replied:
Over here: https://github.com/pytorch/pytorch/blob/6bf0e3b697ce688bc8325440dea3b51fea571c3d/torch/_dynamo/variables/tensor.py#L271. 
Basically when calling item on an input tensor.

voznesenskym replied:
Lol, I mean, what is the case for it? how have we not needed it till now? The answer is "Operations invoked on inputs"

tugsbayasgalan committed:
Update on "[dynamo] Fix bug in tensor.item fake tensor propogation"


When we run the node with fake value for tensor.item, it would previously error because the utility method doesn't know how to handle placeholder node. The tensor we are calling item can be input from user will be placeholder in the graph. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan commented:
@pytorchbot merge -f "flaky test"
